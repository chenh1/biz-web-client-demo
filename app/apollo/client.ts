// import { ApolloLink, HttpLink } from "@apollo/client";
// import { onError } from "@apollo/client/link/error";

// import {
//   NextSSRInMemoryCache,
//   NextSSRApolloClient,
// } from "@apollo/experimental-nextjs-app-support/ssr";
// import { NextSSRInMemoryCache } from "@apollo/experimental-nextjs-app-support/ssr";
// import { registerApolloClient } from "@apollo/experimental-nextjs-app-support";}

// const errorLink = onError(({ graphQLErrors, networkError }) => {
//   if (graphQLErrors)
//     graphQLErrors.forEach(({ message, locations, path, ...rest }) => {
//       Object.keys(rest)?.map(key => console.log(key, rest[key]))
//       console.log(
//         `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
//       )
//     });
//   if (networkError) console.log(`[Network error]: ${networkError}`);
// });

// const httpLink = new HttpLink({
//   uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
// })

// export const { getClient } = registerApolloClient(() => {
//   return new NextSSRApolloClient({
//     cache: new NextSSRInMemoryCache(),
//     link: ApolloLink.from([errorLink, httpLink]),
//   });
// });

import { HttpLink } from "@apollo/client";
import {
  registerApolloClient,
  ApolloClient,
  InMemoryCache,
} from "@apollo/experimental-nextjs-app-support";

export const { getClient, query, PreloadQuery } = registerApolloClient(() => {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      // this needs to be an absolute url, as relative urls cannot be used in SSR
      uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
      // you can disable result caching here if you want to
      // (this does not work if you are rendering your page with `export const dynamic = "force-static"`)
      // fetchOptions: { cache: "no-store" },
    }),
  });
});