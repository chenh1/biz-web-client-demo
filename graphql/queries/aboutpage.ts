import { gql, DocumentNode } from "@apollo/client";

const aboutPageQuery: DocumentNode = gql`
  query AboutPage {
    aboutPage {
      data {
        attributes {
          title
          content
          topImage {
            data {
              attributes {
                url
                width
                height
                alternativeText
              }
            }
          }
        }
      }
    }
  }
`;

export default aboutPageQuery;