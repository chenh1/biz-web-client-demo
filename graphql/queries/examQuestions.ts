import { gql, DocumentNode } from "@apollo/client";

export type MultiPoints = {
  open: number;
  high: number;
  low: number;
  close: number;
}

export type ChartData = {
  multiPoints: MultiPoints;
  x: number;
  y: number;
}

export type Answer = {
  key: string;
  answer: string;
  value?: string;
  xCell?: number;
  yCell?: number;
  x?: number;
  y?: number;
}

export type Cell = {
  header: string;
  position: number;
}

export type CellValue = {
  value: string | number;
  xCell: number;
  yCell: number;
}

export type Question = {
  question: string;
  correctAnswers: Answer[];
  
}

export type InteractiveTableQuestion = Question & {
  answers: Answer[];
  tableData: CellValue[];
  rowHeaders: Cell[];
  columnHeaders: Cell[];
  answerType: string;
}

export type MultiChoiceQuestion = Question & {
  answers: Answer[];
}

export type InteractiveChartQuestion = Question & {
  xType: string;
  yType: string;
  xAxisLabel: string;
  yAxisLabel: string;
  chartType: "line" | "candlestick";
  chartData: ChartData[];
  interactiveChartData: ChartData[];
  correctAnswer: Answer;
}

export type ExamQuestionType = {
  explanation: string;
  lessonModule: number;
  multiChoiceQuestion?: MultiChoiceQuestion;
  fillInQuestion?: Question;
  interactiveTableQuestion?: InteractiveTableQuestion;
  interactiveChartQuestion?: InteractiveChartQuestion;
  tags: {
    data: {
      attributes: {
        displayName: string;
        color: string;
      }
    }
  }
}

export type ExamQuestionResponse = {
  examQuestions: {
    data: {
      attributes: ExamQuestionType
    }[]
  }
}

export type Datum = {
  [key: string]: number | string;
}

export const examQuestionsQuery: DocumentNode = gql`
  query ExamQuestions($lessonModule: Float) {
    examQuestions(filters: {lessonModule: {eq: $lessonModule }}) {
      data {
        attributes {
          tags {
            data {
              attributes {
                displayName
                color
              }
            }
          }
          explanation
          lessonModule
          multiChoiceQuestion {
            question
            answers {
              answer
              key
            }
            correctAnswers {
              key
            }
          }
          fillInQuestion {
            question
            correctAnswers {
              answer
            }
          }
          interactiveTableQuestion {
            question
            answerType
            tableData(pagination: { page: 1, pageSize: 100 }) {
              value
              xCell
              yCell
            }
            correctAnswers {
              value
              xCell
              yCell
            }
            rowHeaders {
              header
              position
            }
            columnHeaders {
              header
              position
            }
          }
          interactiveChartQuestion {
            xType
            yType
            xAxisLabel
            yAxisLabel
            question
            chartType
            chartData(pagination: { page: 1, pageSize: 100 }) {
              x
              y
              multiPoints
            }
            correctAnswer {
              x
              y
            }
          }
        }
      }
    }
  }
`