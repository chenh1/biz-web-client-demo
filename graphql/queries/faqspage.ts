import { gql, DocumentNode } from "@apollo/client";

const faqsPageQuery: DocumentNode = gql`
  query FaqsPage {
    faqsPage {
      data {
        attributes {
          title
          description
          faq(pagination: { page: 1, pageSize: 50 }) {
            question
            answer
            category {
              data {
                attributes {
                  displayName
                  value
                }
              }
            }
            answerRich
          }
        }
      }
    }
  }
`;

export default faqsPageQuery;