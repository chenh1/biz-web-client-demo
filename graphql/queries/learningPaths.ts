import { gql, DocumentNode } from "@apollo/client";

export type RawTree = {
  name: string;
  isCompleted: boolean;
  children?: RawTree[]
}

export const learningPathsQuery: DocumentNode = gql`
  query LearningPaths($level: Int) {
    learningPaths(sort: "level:asc", filters: {level: {eq: $level }}) {
      data {
        attributes {
          title
          level
          description
          rawTree
        }
      }
    }
  }
`

export const allLearningPathsQuery: DocumentNode = gql`
  query LearningPaths {
    learningPaths(sort: "level:asc") {
      data {
        attributes {
          title
          level
          description
          rawTree
        }
      }
    }
  }
`