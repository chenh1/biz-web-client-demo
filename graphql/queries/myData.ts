import { gql, DocumentNode } from "@apollo/client";

export type ProgressData = {
  currentModule?: {
    lessonModule?: string;
  };
  completedModules?: Array<{
    lessonModule: string;
    examScore?: number;
  }>;
}

export type MyDataRes = {
  myData: {
    progressData: ProgressData
  } | {};
}

export const myDataQuery: DocumentNode = gql`
  query MyData {
    myData {
      progressData
    }
  }
`;
