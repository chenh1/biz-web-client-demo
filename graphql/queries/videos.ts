import { gql, DocumentNode } from "@apollo/client";

export type VideoType = {
  title: string;
  url: string;
  description: string;
  transcript: string;
  lessonModule: number | string;
  tags: {
    data: {
      attributes: {
        displayName: string;
        color: string;
        value: string;
      }
    }[]
  }
}

export type Video = {
  attributes: VideoType;
}

export type VideoTypeResponse = {
  videos: {
    data: {
      attributes: VideoType
    }[]
  }
}

export const videosQuery: DocumentNode = gql`
  query Videos($firstModule: Float, $lastModule: Float) {
    videos(pagination: { page: 1, pageSize: 100 }, sort: "lessonModule:asc", filters: {lessonModule: {between: [ $firstModule, $lastModule ] }}) {
      data {
        attributes {
          title
          url
          description
          transcript
          lessonModule
          tags {
            data {
              attributes {
                displayName
                value
                color
              }
            }
          }
        }
      }
    }
  }
`