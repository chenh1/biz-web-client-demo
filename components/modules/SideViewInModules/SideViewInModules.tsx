'use client'

import type { RawTree } from "../../../graphql/queries/learningPaths";
import type { VideoType } from "../../../graphql/queries/videos";
import type { SessionWithJwt } from "../../constants/types";
import React from "react";
import { archetypeIcons } from "../../constants/archetypes";
import { useQuery } from "@apollo/client";
import { Box } from "../../containers/Box/Box";
import { TreeProgress } from "../TreeProgress/TreeProgress";
import { ListProgress } from "../ListProgress/ListProgress";
import { myDataQuery } from "../../../graphql/queries/myData";
import { useSession } from "next-auth/react";
import { allLearningPathsQuery } from "../../../graphql/queries/learningPaths";
import colors from "../../constants/colors";
import { videosQuery } from "../../../graphql/queries/videos";

interface SideViewInModulesProps {
  isVertical: boolean;
  lessonModule: number;
  availableModules: { attributes: VideoType }[]
  rawTree: RawTree
}

export const SideViewInModules: React.FC<SideViewInModulesProps> = ({ isVertical, lessonModule, availableModules, rawTree }: SideViewInModulesProps) => {
  const session = useSession().data as SessionWithJwt | null;

  const { data: { myData: { progressData } = {} } = {} } = useQuery(myDataQuery, { 
    context: { 
      headers: { 
        'Authorization': `Bearer ${session?.data?.jwt}` 
      },
    },
    skip: session?.status !== 'authenticated'
  })

  const learningPathTree = JSON.parse(JSON.stringify(progressData?.customPath || {}))
  const showList = !!progressData
  const archetype = progressData?.archetype

  const { data: { learningPaths: { data: allLearningPaths } = {} } = {} } = useQuery(allLearningPathsQuery)

  const nameMapCallback = (nameMap, nodeName) => {
    return nameMap?.filter(({ attributes: { lessonModule } }) => lessonModule?.toString() === nodeName?.toString())[0]?.attributes?.title
  }

  const firstModule = Math.floor(lessonModule);
  const lastModule = firstModule + 0.999;

  const { data: { videos: { data: videos } = {} } = {} } = useQuery(videosQuery, {
    variables: {
      firstModule: showList ? 0 : firstModule,
      lastModule: showList ? 100 : lastModule
    }
  })
  
  return showList ?
    <Box gap="md">
      {!!archetype &&
        <Box padding="0" className="justify-self-center">
          <img src={archetypeIcons[archetype]} alt="archetype icon" width={120} height={120}/>
        </Box>
      }
      <ListProgress
        isVertical
        nameMapCallback={nameMapCallback}
        nameMap={videos}
        dataCy="current-path"
        // globalOnMouseEnter={(node) => setNodeData(node)}
        progressData={progressData}
        //globalOnMouseLeave={() => setNodeData(null)}
        rawTree={learningPathTree}
        highlightValue={progressData?.currentModule?.lessonModule}
        // className={`col-span-1 w-full min-h-[${nodeCount * 100}px]`}
        background={colors.nightBlack}
        allLearningPaths={allLearningPaths}
        faderColor={colors.nightBlack300}
        availableModules={availableModules}
        showAsList
        nodeWidth={12}
        nodeHeight={12}
      />
    </Box>
    :
    <TreeProgress
      progressData={progressData}
      isVertical={isVertical}
      availableModules={availableModules}
      rawTree={rawTree}
    />
}