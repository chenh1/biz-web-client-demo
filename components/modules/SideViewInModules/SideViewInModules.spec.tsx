import { myDataQuery } from "../../../graphql/queries/myData";
import { allLearningPathsQuery } from "../../../graphql/queries/learningPaths";
import { videosQuery } from "../../../graphql/queries/videos";
import { SideViewInModules } from "./SideViewInModules";
import { MockedProvider } from "@apollo/client/testing";
import { fireEvent, render, waitFor } from '@testing-library/react';
import { JSDOM } from "jsdom"
// import { resetPasswordMutation } from "../../../graphql/mutations/resetPassword";
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

jest.mock('next-auth/react', () => ({
  useSession: jest.fn()
}))
import { useSession } from "next-auth/react";
const mockedUseSession = useSession as jest.Mock;

jest.mock("next/navigation", () => ({
  useRouter() {
    return {
      prefetch: () => null,
      replace: (arg) => null
    };
  },
  useParams() {
    return {
      id: "1"
    }
  },
  useSearchParams: () => {
    return {
      get: () => "123"
    }
  }
}));

jest.mock("@visx/responsive", () => ({
  ParentSize: ({ children }: { children: (size: { width: number, height: number }) => JSX.Element }) => children({ width: 100, height: 100 })
}))

const mocks = [
  {
    request: {
      query: myDataQuery,
    },
    result: {
      data: {
        myData: {
          progressData: {
            currentModule: {
              lessonModule: 1.01
            },
            completedModules: [
              { lessonModule: 1.01, examScore: 100 },
              { lessonModule: 1.02, examScore: 100 }
            ]
          }
        }
      }
    }
  },
  {
    request: {
      query: allLearningPathsQuery
    },
    result: {
      data: {
        learningPaths: {
          data: [
            { attributes: { title: "Intro to cooking", level: "1", description: "Learn the basics of cooking", rawTree: JSON.parse(JSON.stringify({ name: "Lesson 1", description: "Some description", isCompleted: false, children: [ { name: "Lesson 2", description: "Some description", isCompleted: false, children: [] } ] })) } },
            { attributes: { title: "Intermediate cooking", level: "2", description: "Take the next steps", rawTree: JSON.parse(JSON.stringify({ name: "Lesson 1", description: "Some description", isCompleted: false, children: [ { name: "Lesson 2", description: "Some description", isCompleted: false, children: [] } ] })) } },
            { attributes: { title: "Master chef", level: "3", description: "Gain master in cooking", rawTree: JSON.parse(JSON.stringify({ name: "Lesson 1", description: "Some description", isCompleted: false, children: [ { name: "Lesson 2", description: "Some description", isCompleted: false, children: [] } ] })) } }
          ]
        }
      }
    }
  },
  {
    request: {
      query: videosQuery,
    },
    result: {
      data: {
        videos: {
          data: [
            {
              attributes: {
                title: "Title 1",
                url: "https://www.youtube.com/watch?v=1",
                description: "Description 1",
                transcript: "Transcript 1",
                lessonModule: 1.01,
                tags: {
                  data: [
                    {
                      attributes: {
                        displayName: "Tag 1",
                        color: "red",
                        value: "tag1"
                      }
                    }
                  ]
                }
              }
            },
            {
              attributes: {
                title: "Title 2",
                url: "https://www.youtube.com/watch?v=2",
                description: "Description 2",
                transcript: "Transcript 2",
                lessonModule: 1.02,
                tags: {
                  data: [
                    {
                      attributes: {
                        displayName: "Tag 2",
                        color: "blue",
                        value: "tag2"
                      }
                    }
                  ]
                }
              }
            }
          ]
        }
      }
    }
  }
]


it('renders the side view paths module', async () => {
  mockedUseSession.mockImplementation(() => ({ status: "authenticated" }));

  const { getByTestId, container } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <SideViewInModules 
        isVertical={true} 
        lessonModule={1.01}
        availableModules={[
          { attributes: { title: "Intro to cooking", url: "https://www.test.com", description: "test", transcript: "test", lessonModule: 1.01, tags: { data: [{ attributes: { displayName: "Tag 1", color: "blue", value: "tag1" }}] } }},
          { attributes: { title: "Intermediate cooking", url: "https://www.test.com", description: "test", transcript: "test", lessonModule: 1.02, tags: { data: [{ attributes: { displayName: "Tag 1", color: "blue", value: "tag1" }}] } }},
          { attributes: { title: "Master chef", url: "https://www.test.com", description: "test", transcript: "test", lessonModule: 1.03, tags: { data: [{ attributes: { displayName: "Tag 1", color: "blue", value: "tag1" }}] } }}
        ]}
        rawTree={{
          name: "1.01", isCompleted: false, children: [ {
            name: "1.02", isCompleted: false, children: [ {
              name: "1.03", isCompleted: false, children: [] 
            }] 
          }] 
        }}
      />
    </MockedProvider>
  )

  expect(container).toMatchSnapshot();
});
