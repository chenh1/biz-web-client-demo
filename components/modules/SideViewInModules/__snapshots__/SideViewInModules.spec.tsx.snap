// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the side view paths module 1`] = `
<div>
  <div
    class=" grid  gap-4  p-2 overflow-y-auto h-screen w-screen md:w-[50vw]"
  >
    <svg
      height="100"
      width="100"
    >
      <defs>
        <lineargradient
          id="lg"
          x1="0"
          x2="0"
          y1="0"
          y2="1"
        >
          <stop
            offset="0%"
            stop-color="#fd9b93"
            stop-opacity="1"
          />
          <stop
            offset="100%"
            stop-color="#fe6e9e"
            stop-opacity="1"
          />
        </lineargradient>
      </defs>
      <rect
        fill="#272b4d"
        height="100"
        rx="14"
        width="100"
      />
      <g
        class="visx-group"
        transform="translate(80, 80)"
      >
        <path
          class="visx-link visx-link-vertical-diagonal"
          d="M-30,0C-30,-20,-30,-20,-30,-40"
          fill="none"
          stroke="#374469"
          stroke-width="2"
        />
        <path
          class="visx-link visx-link-vertical-diagonal"
          d="M-30,-40C-30,-60,-30,-60,-30,-80"
          fill="none"
          stroke="#374469"
          stroke-width="2"
        />
        <g
          class="visx-group cursor-pointer"
          transform="translate(-30, 0)"
        >
          <rect
            fill="url('#lg')"
            height="56"
            rx="10"
            stroke-dasharray="2,2"
            stroke-opacity="0.6"
            stroke-width="1"
            width="4.8ch"
            x="-2.4ch"
            y="-28"
          />
          <text
            data-cy="-root-text"
            dy=".33em"
            fill="#41ABFF"
            font-family=""Space Grotesk", sans-serif"
            font-size="18"
            style="pointer-events: none; width: 200px;"
            text-anchor="middle"
            width="300px"
          >
            1.01
          </text>
        </g>
        <g
          class="visx-group cursor-pointer relative"
          transform="translate(-30, -40)"
        >
          <circle
            data-cy="-node-1"
            fill="#272b4d"
            height="56"
            r="28"
            stroke="#41ABFF"
            stroke-width="1"
            width="56"
            x="-28"
            y="-28"
          />
          <text
            class=""
            data-cy="-node-text-1"
            dx="0"
            dy=".33em"
            fill="#ffffff"
            font-family=""Space Grotesk", sans-serif"
            font-size="18"
            style="pointer-events: none;"
            text-anchor="middle"
          >
            1.02
          </text>
        </g>
        <g
          class="visx-group cursor-pointer relative"
          transform="translate(-30, -80)"
        >
          <circle
            data-cy="-node-2"
            fill="#272b4d"
            height="56"
            r="28"
            stroke="#41ABFF"
            stroke-width="1"
            width="56"
            x="-28"
            y="-28"
          />
          <text
            class=""
            data-cy="-node-text-2"
            dx="0"
            dy=".33em"
            fill="#ffffff"
            font-family=""Space Grotesk", sans-serif"
            font-size="18"
            style="pointer-events: none;"
            text-anchor="middle"
          >
            1.03
          </text>
        </g>
      </g>
    </svg>
  </div>
</div>
`;
